python/feldera/sql_context.py

```diff
@@ -0,0 +1,278 @@
+
+        name = _table_name_from_sql(ddl)
+
+        self.tables[name] = SQLTable(name, ddl)
+
+    def input_from_pandas(self, table_name: str, df: pandas.DataFrame):
+        """
+        Adds a pandas DataFrame to the input buffer of the SQLContext, to be pushed to the pipeline.
+
+        :param table_name: The name of the table.
+        :param df: The pandas DataFrame to be pushed to the pipeline.
+        """
+
+        tbl = self.tables.get(table_name)
+
+        if tbl:
+            # tbl.validate_schema(df)   TODO: something like this would be nice
+            self.http_input_buffer.append({tbl.name: df.to_dict('records')})
+            return
+
+        tbl = self.todo_tables.get(table_name)
+
+        if not tbl:
+            raise ValueError(f"Table {table_name} not registered")
+
+        # tbl.infer_schema(df)  TODO: support schema inference
+
+        self.tables[table_name] = tbl
+        self.todo_tables.pop(table_name)
+
+        self.http_input_buffer.append({tbl.name: df.to_dict('records')})
+
+    def register_view(self, name: str, query: str):
+        """
+        Registers a Feldera View based on the provided query.
+        Auto inserts the trailing semicolon if not present.
+
+        :param name: The name of the view.
+        :param query: The query to be used to create the view.
+        """
+
+        if query[-1] != ';':
+            query += ';'
+
+        self.views[name] = f"CREATE VIEW {name} AS {query}"
+
+    def listen(self, view_name: str) -> OutputHandler:
+        """
+        Listens to the output of the provided view so that it is available in the notebook / python code.
+
+        :param view_name: The name of the view to listen to.
+        """
+
+        queue = Queue(maxsize=1)
+
+        self.views_tx.append({view_name: queue})
+
+        handler = OutputHandler(self.client, self.pipeline_name, view_name, queue)
+        handler.start()
+
+        return handler
+
+    def run_to_completion(self):
+        """
+        Runs the pipeline to completion, waiting for all input records to be processed.
+
+        :raises RuntimeError: If the pipeline returns unknown metrics.
+        """
+
+        self.__setup()
+
+        self.client.start_pipeline(self.pipeline_name)
+
+        for view_queue in self.views_tx:
+            for view_name, queue in view_queue.items():
+                queue.put(_OutputHandlerInstruction.PipelineStarted)
+                queue.join()
+
+        for input_buffer in self.http_input_buffer:
+            for tbl_name, data in input_buffer.items():
+                self.client.push_to_pipeline(self.pipeline_name, tbl_name, "json", data, array=True)
+
+        while True:
+            metrics: dict = self.client.get_pipeline_stats(self.pipeline_name).get("global_metrics")
+            input_records = metrics.get("total_input_records")
+            processed_records = metrics.get("total_processed_records")
+
+            if input_records == processed_records:
```

Comment:
This is not a correct way to check that the pipeline has run to completion. Use the `pipeline_complete` flag instead.